universe/NamedValueRefManager.cpp

```diff
@@ -0,0 +1,190 @@
+    std::transform(m_value_refs.begin(), m_value_refs.end(), std::inserter(aet, aet.end()), base_to_base);
+    return aet;
+}
+
+template <typename R, typename VR>
+void NamedValueRefManager::RegisterValueRefImpl(R& container, std::mutex& mutex, const std::string& label, std::string&& valueref_name, std::unique_ptr<VR>&& vref) {
+    InfoLogger() << "Register " << label << " valueref for " << valueref_name << ": " << vref->Description();
+    if (container.count(valueref_name)>0) {
+        DebugLogger() << "Skip registration for already registered " << label << " valueref for " << valueref_name;
+        DebugLogger() << "Number of registered " << label << " ValueRefs: " << m_value_refs.size();
+        return;
+    }
+    const std::lock_guard<std::mutex> lock(mutex);
+    if (!(vref->RootCandidateInvariant() && vref->LocalCandidateInvariant() && vref->TargetInvariant() && vref->SourceInvariant()))
+            ErrorLogger() << "Currently only invariant value refs can be named. " << valueref_name;
+    container.emplace(std::move(valueref_name), std::move(vref));
+    DebugLogger() << "Number of registered " << label << " ValueRefs: " << container.size();
+}
+
+template <typename T>
+void NamedValueRefManager::RegisterValueRef(std::string&& valueref_name, std::unique_ptr<ValueRef::ValueRef<T>>&& vref) {
+    this->RegisterValueRefImpl(m_value_refs, m_value_refs_mutex, "generic", std::move(valueref_name), std::move(vref));
+}
+
+// specialisation for registering to the ValueRef<int> registry
+template <>
+void NamedValueRefManager::RegisterValueRef(std::string&& valueref_name, std::unique_ptr<ValueRef::ValueRef<int>>&& vref) {
+    this->RegisterValueRefImpl(m_value_refs_int, m_value_refs_int_mutex, "int", std::move(valueref_name), std::move(vref));
+}
+
+// specialisation for registering to the ValueRef<double> registry
+template <>
+void NamedValueRefManager::RegisterValueRef(std::string&& valueref_name, std::unique_ptr<ValueRef::ValueRef<double>>&& vref) {
+    this->RegisterValueRefImpl(m_value_refs_double, m_value_refs_double_mutex, "double", std::move(valueref_name), std::move(vref));
+}
+
+NamedValueRefManager& GetNamedValueRefManager()
+{ return NamedValueRefManager::GetNamedValueRefManager(); }
+
+ValueRef::ValueRefBase* const GetValueRefBase(const std::string& name)
+{
+    InfoLogger() << "NamedValueRefManager::GetValueRefBase look for registered valueref for \"" << name << '"';
+    if ( GetNamedValueRefManager().GetValueRefBase(name) )
+        return GetNamedValueRefManager().GetValueRefBase(name);
+    InfoLogger() << "NamedValueRefManager::GetValueRefBase could not find registered valueref for \"" << name << '"';
+    return nullptr;
+}
+
+template <typename T>
+ValueRef::ValueRef<T>* const GetValueRef(const std::string& name)
+{ return GetNamedValueRefManager().GetValueRef<T>(name); }
+
+template <typename T>
+void RegisterValueRef(std::string name, std::unique_ptr<ValueRef::ValueRef<T>>&& vref)
```

Comment:
could perfect forward `name` since it's already a templated function

Reply:
not sure what you mean/how that works exactly/how i can declare ::RegisterValueRef in ValueRef.h then. 

the following would make sense in the definition (because that uses name as a string) but from the declaration in the header no use site would know that this is about strings:
```
template <typename S, typename T>
void RegisterValueRef(S&& name, std::unique_ptr<ValueRef::ValueRef<T>&& vref)
```

Reply:
Hmm, I'm a bit confused by the multiple layers of valueref registering functions... seems unnecessarily complex and not really making good use of the templating...